src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.c

```diff
@@ -1611,13 +1698,30 @@ static void lb_on_response_received_locked(grpc_exec_ctx *exec_ctx, void *arg,
              * update or in glb_destroy() */
             glb_policy->serverlist = serverlist;
             glb_policy->serverlist_index = 0;
+            /* dispose the fallback (if any). It should only be done once when the 
```

Comment:
I think using `fallback_backend_addresses` to determine the fallback status is a little better. And my main reason is that it's more coupled (inaccurate word, but I can't find a better one) with the newly-added fallback code and less coupled with the existing `serverlist` code. 

Specifically, `fallback_backend_addresses` is exclusively for the fallback logic, using it depends less on the other code and reduces the risk that we need to change it if the depended code changes in the future (e.g., the expiration is used). Also, we won't mix the code dealing with the `serverlist` (from line 1692 to line 1700) and the code dealing with the fallback (from line 1701 to line 1711), which I think is good to have. (Plus the order looks good to me: "hey the new server list is ready in my hand now", "great, let's throw away the fallback". Although flipping them doesn't introduce any issue to our code...)

Many of the reasons are quite weak or/and personal, but I want to clarify them by my best effort. The only advantage I've understood of using `serverlist` instead is that we can reuse the if clause, which I think is even less important. However, if the advice is still to change it to use `serverlist`, I will accept it and I believe there is a reason for that but I haven't caught successfully.

Reply:
I think there are two issues here.

First, it seems that you're very concerned about the possibility of needing to handle the expiration case in the future, and I don't think that's something we need to worry about.  What we discussed at the meeting last week was that we were going to remove the expiration field from the load balancer protocol, because its behavior is not actually specified and it's not at all clear what it should do.  We have no plans to support this in the future, and if we did, we would need to write a new spec from scratch as to what its behavior would be.  Given that this is completely speculative, I don't think this is something we should plan for.

Second, I disagree that the fallback code is independent of the serverlist.  By definition, the fallback is what happens when there's something wrong the serverlist.  Currently, "there's something wrong the serverlist" is defined as `serverlist == NULL`.  It's certainly possible that that will change in the future; for example, if we decide to use the fallback even after we get a serverlist if we can no longer reach any of the backends listed in the serverlist, then we might change  `serverlist == NULL` to something like `serverlist == NULL || serverlist->num_reachable_servers == 0`.  But either way, it would still be the case that the decision about whether to use the fallback will be based on the serverlist.  Ultimately, the serverlist is the primary source of information, and the fallback is what we use if there's something wrong with that primary source of information.

With regard to the ordering, as I've said before, the locking here means that that's really not an issue.  And I think that as you gain more experience with our codebase, you'll see that there are many places where this sort of thing happens, and that it's something you will get used to.

Reply:
I see. I think I know how we viewed this differently now. I agree we should use `serverlist`.